pylint/checkers/classes.py

```diff
@@ -1308,8 +1309,74 @@ def _check_protected_attribute_access(self, node):
                     if _is_attribute_property(name, klass):
                         return
 
+                #  A licit use of protected member is inside a special method
+                #  if the expression that use this member is the same class as
+                #  the current one
+                if (
+                    not attrname.startswith("__")
+                    and self._is_called_inside_special_method(node)
+                    and self._is_instance_of_current_class(node.expr)
+                ):
+                    return
+
                 self.add_message("protected-access", node=node, args=attrname)
 
+    @staticmethod
+    def _is_called_inside_special_method(node: astroid.node_classes.NodeNG) -> bool:
+        """
+        Returns true if the node is located inside a special (aka dunder) method
+        """
+        try:
+            frame_name = node.frame().name
+        except AttributeError:
+            return False
+        return frame_name and frame_name in PYMETHODS
+
+    @staticmethod
+    def _get_if_nodes_parent(
+        node: astroid.node_classes.NodeNG
+    ) -> Optional[List[astroid.node_classes.If]]:
+        """
+        Returns the list of If nodes that are parents of the node
+        """
+        res = []
+        if_node = node.parent
+        while if_node is not None:
+            if isinstance(if_node, astroid.node_classes.If):
+                res.append(if_node)
+            if_node = if_node.parent
+        return res
+
+    @staticmethod
+    def _is_instance_of_current_class(node: astroid.node_classes.Name) -> bool:
+        """
+        Returs true if the node is the name of an instance of the same class
+
+        :note: test that at least a call to 'isinstance(node, self.__class__)' is made
+               in the parent If nodes
+        """
+        if_nodes_parent = ClassChecker._get_if_nodes_parent(node)
+        if not if_nodes_parent:
+            return False
+
+        # Check if one of the test made in the parent If nodes ensures that the node
+        #  is an instance of the current class.
+        for if_node in if_nodes_parent:
```

Comment:
@hippo91 I find this extra unnecessary complexity 😅 The implementation would be a lot simpler if we would just skip emitting `protected-access` inside a special method, assuming the person that wrote the code already handled the case of accessing the attribute on a same instance.